Add link to analytics options in submenu and make all settings sections linkable

johnwatkins0 opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Fixes #5190  

In #5155 we removed the Analytics menu page and placed the analytics option on the settings screen. This restores that menu item with a link to the Analytics section of the menu page.

Additional changes:
- Passed the URL hash into the root settings screen component so that it can be used to scroll to sections after the UI loads
- Moved the drawers on the settings screen up to the root settings screen component so that all the handling of IDs/anchors can be in one file
- Created an `AnalyticsOptionsSubmenu` service class, which is a stripped-down version of the class that was removed in #5155. It inserts a link into the AMP admin menu in the second position. 
- Created a `get_menu_slug` method in the `OptionsMenu` class for consumption in `AnalyticsOptionsSubmenu`. 

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
Add link to analytics options in submenu and make all settings sections linkable

johnwatkins0 committed:
Fix tests for options menu

johnwatkins0 committed:
PHPUnit test: adjust expected menu item count

johnwatkins0 committed:
PHPUnit: restore expected link for user with dev tools on

johnwatkins0 committed:
Add e2e test for settings page anchor linking

johnwatkins0 committed:
Remove unused imports

johnwatkins0 committed:
Actually change dev tools setting in phpunit test

johnwatkins0 committed:
Remove unnecessary use statement

johnwatkins0 committed:
Remove unnecessary space

johnwatkins0 committed:
Comment fixes

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/5190-settings-screen-anchor-links

* 'develop' of github.com:ampproject/amp-wp:
  Revert "Replace astra with neve for testing since astra is suspended"
  Unset `$_SERVER['REQUEST_URI']`
  Fix lint error
  Default to localhost if `$_SERVER['HTTP_HOST']` is not set
  Remove check for WP CLI context
  Default parsed URL host to localhost if in WP CLI context
  Default parsed URL host to localhost if in WP CLI context
  Add null as return type
  Return null if the URL cannot be parsed when in a WP CLI environment
  Padding: adjust padding on mobile-width close button
  Ensure non-AMP link in paired browsing does not redirect to AMP version
  Update dependency @wordpress/data to v4.22.3
  Update ReaderThemes::get_themes() to remove any themes that lack slug
  Also make name optional, falling back to slug
  Make theme description optional for Reader themes
  Make homepage a optional field for Reader themes
  Wizard: move close button to the top in mobile view
  Defer mobile redirection to end of template_redirect action
  Fix test referencing core theme
  Abort post-processing if an expected template was not rendered

westonruter commented on assets/src/settings-page/index.js:
```diff
@@ -94,8 +95,11 @@ ErrorNotice.propTypes = {
  * Settings page application root.
+ *
+ * @param {Object} props Component props.
+ * @param {string} props.section The initially focused section.
```
Maybe `focusedSection` would be more clear?

johnwatkins0 replied:
Done in e41c971d9

westonruter commented on assets/src/settings-page/index.js:
```diff
@@ -120,6 +124,25 @@ function Root() {
+		const focusedSection = document.getElementById( section );
+
+		if ( focusedSection ) {
+			focusedSection.scrollIntoView();
```
Would smooth scrolling be desirable? 

```suggestion
			focusedSection.scrollIntoView( { behavior: 'smooth' } );
```

westonruter replied:
Alternatively, for accessibility, I think what is needed is to actually `focus()` on the first focusable element. 

So this should be looking for the first `button` or `input` that is inside the `focusedSection` and then `focus()` on it. This will automatically scroll _and_ add the necessary keyboard focus.

johnwatkins0 replied:
Done in `e41c971d9` with `select` elements also included

westonruter commented on src/Admin/AnalyticsOptionsSubmenu.php:
```diff
@@ -0,0 +1,73 @@
+			[ $old_menu[0] ],
+			[
+				[
+					__( 'AMP Analytics Options', 'amp' ),
```
This needs to be shortened to just `Analytics`:

![image](https://user-images.githubusercontent.com/134745/90057305-0e842300-dc95-11ea-870b-17a49a13edd5.png)

johnwatkins0 replied:
Done in b28aa90a4

westonruter commented on src/Admin/AnalyticsOptionsSubmenu.php:
```diff
@@ -0,0 +1,73 @@
+			array_slice( $old_menu, 1 )
+		);
+
+		$submenu[ $this->parent_menu_slug ] = array_values( $new_menu ); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
```
Should not `$submenu` be modified with `add_submenu_page()` rather than doing this and the above?

johnwatkins0 replied:
Ah, you're right. Updated in 0299a03c0. Didn't realize it would be so simple to link to the parent menu page with a hash appended.

westonruter commented on assets/src/settings-page/index.js:
```diff
@@ -161,7 +222,7 @@ domReady( () => {
 			<ErrorContextProvider>
 				<Providers>
-					<Root optionsRestPath={ OPTIONS_REST_PATH } />
+					<Root section={ global.location.hash.replace( /^#/, '' ) } />
```
I think this can be improved… namely, if I am already on the AMP Settings screen and click the admin menu item, then I am jumped down the screen without the section expanding. Therefore, it would be preferable if there was also a [`hashchange`](https://developer.mozilla.org/en-US/docs/Web/API/Window/hashchange_event) event that would re-render this tree whenever it fires. This will ensure that the section always reflects what is currently in the `location.hash`.

johnwatkins0 replied:
I did this in 754aadd8572de9744af0f99953b077c3bded9bf9 but it was surprisingly complicated to get it to work in all the scenarios (initial load of the page, changing the hash in the URL location bar, clicking the Analytics link in the menu), and I'm not sure how I feel about how much code I had to add. It's working, though.

johnwatkins0 committed:
Add menu item with add_submnu_page

johnwatkins0 committed:
Merge branch 'feature/5190-settings-screen-anchor-links' of https://github.com/ampproject/amp-wp into feature/5190-settings-screen-anchor-links

johnwatkins0 committed:
Shorten Analytics menu link name

johnwatkins0 committed:
Settings anchor linking: smoothly scroll to first focusable element in linked section

westonruter commented on assets/src/settings-page/index.js:
```diff
@@ -94,8 +95,13 @@ ErrorNotice.propTypes = {
 
 /**
  * Settings page application root.
+ *
+ * @param {Object} props Component props.
+ * @param {string} props.focusedSection The initially focused focusedSection.
```
```suggestion
```

westonruter commented on assets/src/settings-page/index.js:
```diff
@@ -153,6 +221,9 @@ function Root() {
 		</>
 	);
 }
+Root.propTypes = {
+	focusedSection: PropTypes.string,
+};
```
```suggestion
```

westonruter commented on assets/src/settings-page/index.js:
```diff
@@ -94,8 +95,13 @@ ErrorNotice.propTypes = {
+ * @param {string} props.focusedSection The initially focused focusedSection.
  */
 function Root() {
+	const [ focusedSection, setFocusedSection ] = useState( global.location.hash.replace( /^#/, '' ) );
```
`setFocusedSection` isn't yet used. It will be set in a `hashchange` event handler?

johnwatkins0 replied:
Yup, in 754aadd8572de9744af0f99953b077c3bded9bf9

johnwatkins0 committed:
Update handling of scroll and hash behavior on settings screen

johnwatkins0 committed:
Update how drawer initialOpen resets (fixing Jest test)

johnwatkins0 committed:
Resolve small failures in PHPUnit tests

pierlon commented on assets/src/settings-page/index.js:
```diff
@@ -120,6 +144,41 @@ function Root() {
+				event.preventDefault();
+			}
+
+			// Ensure this runs after state upates.
```
```suggestion
			// Ensure this runs after state updates.
```

pierlon commented on assets/src/settings-page/index.js:
```diff
@@ -132,13 +191,48 @@ function Root() {
+				<h2 id="advanced-settings">
 					{ __( 'Advanced Settings', 'amp' ) }
 				</h2>
 				<MobileRedirection />
```
Why not put the ID for the section here instead of inside of the component?

johnwatkins0 replied:
Yeah, I like having all the IDs at this level. Updated in 5e7b5de0734c92a5ab0f163656ee6abf21af96e6

pierlon commented on assets/src/settings-page/index.js:
```diff
@@ -132,13 +191,48 @@ function Root() {
+					id="supported-templates"
+					initialOpen={ 'supported-templates' === focusedSection }
+				>
+					<SupportedTemplates />
```
Same question as above.

johnwatkins0 replied:
In this case the ID is set just above on the `AMPDrawer` component.

pierlon commented on tests/e2e/utils/visit-admin-page-with-hash.js:
```diff
@@ -0,0 +1,35 @@
+import { getPageError } from '@wordpress/e2e-test-utils';
+
+function createURLWithHash( WPPath, query = '', hash = '' ) {
+	const url = new URL( 'http://localhost:8890' );
```
This URL should be available via the environment variable [`WP_BASE_URL`](https://github.com/ampproject/amp-wp/blob/6f92f315964cb5b6629edff2893968bf4ae27b9b/package.json#L131), so no need to hard code it.

johnwatkins0 replied:
Updated in 92ec1f3e6

pierlon commented on tests/e2e/utils/visit-admin-page-with-hash.js:
```diff
@@ -0,0 +1,35 @@
+ * @param {string} query String to be serialized as query portion of URL.
+ * @param {string} hash URL hash.
+ */
+export async function visitAdminPageWithHash( adminPath, query, hash = null ) {
```
If defaulting `hash` to `null` is still wanted I'd recommend updating the docblock to include `null`, otherwise I'd recommend defaulting to an empty string:

```suggestion
export async function visitAdminPageWithHash( adminPath, query, hash = '' ) {
```

johnwatkins0 replied:
Updated in 92ec1f3e6

pierlon commented on tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php:
```diff
@@ -0,0 +1,100 @@
+
+namespace AmpProject\AmpWP\Tests\Admin;
+
+use AMP_Options_Manager;
```
Not being used.

```suggestion
```

pierlon commented on tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php:
```diff
@@ -0,0 +1,100 @@
+/**
+ * Tests for AnalyticsOptionsSubmenu.
+ *
+ * @group options-menu
```
I've recently noticed that the Travis job that runs the PHP tests with code coverage has been emitting a [lot of warnings](https://travis-ci.org/github/ampproject/amp-wp/jobs/716788170#L847), most of which are due to incorrect uses of the `@covers` and `@uses` annotations. I've managed to fix them in https://github.com/ampproject/amp-wp/pull/5201/commits/cfa46c486453e39ec330371fe60944dccdb57f17 but I'd like to prevent them from occurring in the future, so let's start using them the way they're meant to be :smiley:.

/cc @westonruter.

```suggestion
 * @group options-menu
 * @coversDefaultClass \AmpProject\AmpWP\Admin\AnalyticsOptionsSubmenu
```

pierlon commented on tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php:
```diff
@@ -0,0 +1,100 @@
+	/**
+	 * Test register.
+	 *
+	 * @see AnalyticsOptionsSubmenu::register()
```
```suggestion
	 * @covers :register()
```

pierlon commented on tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php:
```diff
@@ -0,0 +1,100 @@
+	/**
+	 * Test add_submenu_link.
+	 *
+	 * @covers AnalyticsOptionsSubmenu::add_submenu_link()
```
```suggestion
	 * @covers ::add_submenu_link()
```

johnwatkins0 committed:
Settings screen: move mobile redirection html ID up to the root component

johnwatkins0 committed:
PHPUnit docblock improvements

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Use ENV variable in e2e tests instead of hardcode URL

johnwatkins0 committed:
Comment fix

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/5190-settings-screen-anchor-links

johnwatkins0 committed:
E2E tests: disable smooth scroll CSS rule for tests where it's relevant

johnwatkins0 committed:
Merge branch 'feature/5190-settings-screen-anchor-links' of https://github.com/ampproject/amp-wp into feature/5190-settings-screen-anchor-links

johnwatkins0 commented on tests/e2e/specs/admin/reader-theme-carousel.js:
```diff
@@ -6,6 +6,7 @@ import { visitAdminPage } from '@wordpress/e2e-test-utils';
 describe( 'AMP settings screen reader themes carousel', () => {
 	beforeEach( async () => {
 		await visitAdminPage( 'admin.php', 'page=amp-options' );
+		await page.addStyleTag( { content: 'html {scroll-behavior: auto !important;}' } );
```
Figuring this out was rough. Puppeteer kept failing with "Node is either not visible or not an HTMLElement", and I found this buried as the last answer in one of the lower-ranking StackOverflow results. Essentially, the `scroll-behavior: smooth` CSS rule I added to the setting screen messes up Puppeteer.

westonruter replied:
Good find.

johnwatkins0 committed:
PHPUnit: menu test updates

johnwatkins0 committed:
PHPUnit: update expected menu item count

johnwatkins0 commented:
Re-requesting review after addressing feedback and getting tests to pass.

johnwatkins0 committed:
Settings: remove unnecessary setTimeout around scrolling newly focused section into view

johnwatkins0 committed:
Remove function unnecessarily returned from effect hook

jwold commented:
I like the change. It took me a moment to catch on that the analytics menu was a jump down, but the rest of the items were separate pages. Apart from that I think it's a good example of still making Analytics findable, but keeping it inside the main settings page. 

![Screen Capture on 2020-08-13 at 13-11-07](https://user-images.githubusercontent.com/3411173/90182314-9f76ff00-dd66-11ea-964c-faf6d09e3c50.gif)

johnwatkins0 committed:
Polyfill wp-components CSS on Amp screens if wp version is below 5.4

johnwatkins0 committed:
Remove carousel button right margin

johnwatkins0 commented on src/Admin/Polyfills.php:
```diff
@@ -128,6 +128,10 @@ public function register_shimmed_scripts( $wp_scripts ) {
 	 * @param WP_Styles $wp_styles The WP_Styles instance for the current page.
 	 */
 	public function register_shimmed_styles( $wp_styles ) {
+		if ( version_compare( get_bloginfo( 'version' ), '5.4', '<' ) ) {
```
Unrelated change: Noticed some undesirable CSS effects when working in 5.2 locally, so this will force the polyfilled wp-components styles. Note this only runs on AMP screens.

pierlon committed:
Restore the global admin menu when done testing

pierlon commented:
@googlebot I consent.

pierlon commented:
Some tests were modifying the global `$submenu` without restoring it's original value, so I've fixed that in https://github.com/ampproject/amp-wp/pull/5216/commits/70e517b6bf3f5d82c95c0c745d8820d9de3eaed3. I've also updated `OptionsMenuTest::test_add_menu_items` to perform the correct assertions.

pierlon commented:
On a larger viewport the scroll to the Analytics section doesn't seem to have much have an effect:


![](https://user-images.githubusercontent.com/16200219/90211502-62e7e980-dde0-11ea-98ee-f0311b90080c.gif)

johnwatkins0 commented:
> On a larger viewport the scroll to the Analytics section doesn't seem to have much have an effect:


That's true, but it's just because the analytics section is last on the page so it can't scroll any further than that. I suppose the workarounds would be to move that section up or add a bit of extra padding to the bottom of the page.

pierlon commented:
Could we focus the drawer when selected by hash? Similar to how GitHub does it, where if I [link to a comment](https://github.com/ampproject/amp-wp/pull/5216#issuecomment-673900209) it adds a border around it:

![image](https://user-images.githubusercontent.com/16200219/90218620-cc252800-ddf3-11ea-866e-dd83b7fd07fc.png)

westonruter commented:
> Could we focus the drawer when selected by hash? Similar to how GitHub does it, where if I [link to a comment](https://github.com/ampproject/amp-wp/pull/5216#issuecomment-673900209) it adds a border around it:

I'd say that would be an optional enhancement. GitHub styles this via:

```css
:target .timeline-comment {
    border-color: #2188ff;
    box-shadow: 0 0 0 0.2em #c8e1ff;
}
```

However, the drawers depend on `box-shadow` to achieve their effect. If we do any such highlighting, I'd say perhaps we go with an initial faint yellow that fades out (e.g. via a CSS animation).

But yeah, I'd say this is not really necessary. After all, the first focusable element will have the outline. I suppose the the alternative would be remove the `:not(.components-panel__body-toggle)` so that the drawer button gets the border which would have a similar effect to what GitHub is doing:

![image](https://user-images.githubusercontent.com/134745/90220809-c14fa080-ddbd-11ea-8ee0-32606c69ad3c.png)

```diff
--- a/assets/src/settings-page/index.js
+++ b/assets/src/settings-page/index.js
@@ -105,7 +105,7 @@ function scrollFocusedSectionIntoView( focusedSectionId ) {
 		return;
 	}
 
-	const firstInput = focusedSectionElement.querySelector( 'input, select, textarea, button:not(.components-panel__body-toggle)' );
+	const firstInput = focusedSectionElement.querySelector( 'input, select, textarea, button' );
 
 	if ( firstInput ) {
 		firstInput.focus();
```

westonruter commented:
I noticed that when going to `/wp-admin/admin.php?page=amp-options#reader-themes-drawer` it doesn't cause the Reader theme drawer to open initially.

johnwatkins0 committed:
Modifications to section scrollTo behavior on settings screen

johnwatkins0 commented:
a856c71c5 adds scrolling to the reader themes drawer if it's on the page (note I changed that HTML ID from `reader-themes-drawer` to `reader-themes`). I also removed that `:not(.components-panel__body-toggle)` in the selector, so the toggle button will always be focused for these drawers.

Another thing, I noticed that `element.scrollIntoView({behavior: smooth})` seemed to work much more reliably than `focusElement.focus({behavior: smooth})` so I updated it to use `scrollIntoView` on the main element and then focus the focus element without smooth scroll.

westonruter commented:
> `focusElement.focus({behavior: smooth})`

Oh, I didn't even the `focus()` method as supporting `{behavior:smooth}`: https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/focus

So, good to switch to `scrollIntoView` 👍

westonruter commented on assets/src/settings-page/style.css:
```diff
@@ -361,13 +364,13 @@ li.error-kept {
 }
 
-#reader-themes-drawer {
+#reader-themes {
```
Thanks for making this change. I was going to suggest it but I didn't want to nit pick… too much. 😊

johnwatkins0 commented:
> Oh, I didn't even the `focus()` method as supporting `{behavior:smooth}`:

Huh, you're right. I could've sworn I learned this was a thing recently, but I'm no longer able to find it 🤷‍♂️  No wonder it didn't work.

westonruter commented:
For some reason deeplinking to `#reader-themes` is not causing scrolling to occur. If I try navigating in a new tab to `#analytics` I get scrolled down, but not if I go to `#reader-themes`. If, however, I put the scroll logic in a 100ms timeout then the Reader Themes will scroll into view. It's strange because the Reader Themes section does exist in the DOM and it is being rendered. Not a blocker since this section is near the top anyway and it is still getting expanded, but it is still a mystery.

westonruter committed:
Skip gettingElementById for empty strings

westonruter committed:
Remove redundant behavior:smooth from scrollIntoView since scroll-behavior:smooth CSS present
